src/parquet/file/writer.h

```diff
@@ -0,0 +1,171 @@
+
+#include "parquet/schema/descriptor.h"
+#include "parquet/schema/types.h"
+#include "parquet/util/mem-allocator.h"
+
+namespace parquet {
+
+class ColumnWriter;
+class PageWriter;
+class OutputStream;
+
+class RowGroupWriter {
+ public:
+  // Forward declare the PIMPL
+  struct Contents {
+    virtual int num_columns() const = 0;
+    virtual int64_t num_rows() const = 0;
+
+    // TODO: PARQUET-579
+    // virtual void WriteRowGroupStatitics();
+    virtual ColumnWriter* NextColumn() = 0;
+    virtual void Close() = 0;
+
+    // Return const-poitner to make it clear that this object is not to be copied
+    virtual const SchemaDescriptor* schema() const = 0;
+  };
+
+  RowGroupWriter(std::unique_ptr<Contents> contents, MemoryAllocator* allocator);
+
+  /**
+   * Construct a ColumnWriter for the indicated row group-relative column.
+   *
+   * Ownership is solely within the RowGroupWriter. The ColumnWriter is only valid
+   * until the next call to NextColumn or Close. As the contents are diretly written to
+   * the sink, once a new column is started, the contents of the previous one cannot be
+   * modified anymore.
+   */
+  ColumnWriter* NextColumn();
+  void Close();
+
+  int num_columns() const;
+
+  /**
+   * Number of rows that shall be written as part of this RowGroup.
+   */
+  int64_t num_rows() const;
+
+  // TODO: PARQUET-579
+  // virtual void WriteRowGroupStatitics();
+
+ private:
+  // Owned by the parent ParquetFileReader
+  const SchemaDescriptor* schema_;
+
+  // PIMPL idiom
+  // This is declared in the .cc file so that we can hide compiled Thrift
+  // headers from the public API and also more easily create test fixtures.
+  std::unique_ptr<Contents> contents_;
+
+  MemoryAllocator* allocator_;
+};
+
+class ParquetFileWriter {
+ public:
+  // Forward declare the PIMPL
+  struct Contents {
+    virtual ~Contents() {}
+    // Perform any cleanup associated with the file contents
+    virtual void Close() = 0;
+
+    virtual RowGroupWriter* AppendRowGroup(int64_t num_rows) = 0;
+
+    virtual int64_t num_rows() const = 0;
+    virtual int num_columns() const = 0;
+    virtual int num_row_groups() const = 0;
+
+    // Return const-poitner to make it clear that this object is not to be copied
+    const SchemaDescriptor* schema() const {
+       return &schema_;
+    }
+    SchemaDescriptor schema_;
+  };
+
+  ParquetFileWriter();
+  ~ParquetFileWriter();
+
+  static std::unique_ptr<ParquetFileWriter> Open(
+      std::shared_ptr<OutputStream> sink,
+      std::shared_ptr<schema::GroupNode>& schema,
+      MemoryAllocator* allocator = default_allocator());
+
+  void Open(std::unique_ptr<Contents> contents);
+  void Close();
+
+  /**
+   * Construct a RowGroupWriter for the indicated number of rows.
+   *
+   * Ownership is solely within the ParquetFileWriter. The RowGroupWriter is only valid
+   * until the next call to AppendRowGroup or Close.
+   *
+   * @param num_rows The number of rows that are stored in the new RowGroup
```

Comment:
Is there a reason to use C-style comments for Doxygen? Versus C++-style comments (at least in emacs c++-mode, it's easier to always use `//`)